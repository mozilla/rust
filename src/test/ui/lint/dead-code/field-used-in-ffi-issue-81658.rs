//! The field `items` is being "used" by FFI (implicitly through pointers). However, since rustc
//! doesn't know how to detect that, we produce a message that says the field is unused. This can
//! cause some confusion and we want to make sure our diagnostics help as much as they can.
//!
//! Issue: https://github.com/rust-lang/rust/issues/81658

#![deny(dead_code)]

/// A struct for holding on to data while it is being used in our FFI code
pub struct FFIData<T> {
    /// These values cannot be dropped while the pointers to each item
    /// are still in use
    items: Option<Vec<T>>, //~ ERROR field is never read
}

impl<T> FFIData<T> {
    pub fn new() -> Self {
        Self {items: None}
    }

    /// Load items into this type and return pointers to each item that can
    /// be passed to FFI
    pub fn load(&mut self, items: Vec<T>) -> Vec<*const T> {
        let ptrs = items.iter().map(|item| item as *const _).collect();

        self.items = Some(items);

        ptrs
    }
}

extern {
    /// The FFI code that uses items
    fn process_item(item: *const i32);
}

fn main() {
    // Data cannot be dropped until the end of this scope or else the items
    // will be dropped before they are processed
    let mut data = FFIData::new();

    let ptrs = data.load(vec![1, 2, 3, 4, 5]);

    for ptr in ptrs {
        // Safety: This pointer is valid as long as the arena is in scope
        unsafe { process_item(ptr); }
    }

    // Items will be safely freed at the end of this scope
}
