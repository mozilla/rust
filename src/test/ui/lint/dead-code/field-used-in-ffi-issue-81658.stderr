error: field is never read: `items`
  --> $DIR/field-used-in-ffi-issue-81658.rs:13:5
   |
LL |     items: Option<Vec<T>>,
   |     -----^^^^^^^^^^^^^^^^
   |     |
   |     help: if this is intentional, prefix it with an underscore: `_items`
   |
   = note: the leading underscore signals that this field serves some other purpose even if it isn't used in a way that we can detect. (e.g. for its effect when dropped or in foreign code)
note: the lint level is defined here
  --> $DIR/field-used-in-ffi-issue-81658.rs:7:9
   |
LL | #![deny(dead_code)]
   |         ^^^^^^^^^

error: aborting due to previous error

